tests/kernel/mutex/mutex_error_case/src/test_mutex_error.c

```diff
@@ -0,0 +1,251 @@
+	/* User threads aren't allowed to induce kernel panics; generate
+	 * an oops instead.
+	 */
+		if (_is_user_context()) {
+			k_oops();
+		}
+#endif
+		k_panic();
+	}
+}
+/* end of self-defined assert handler */
+
+static void tIsr_entry_lock(const void *p)
+{
+	k_mutex_lock((struct k_mutex *)p, K_NO_WAIT);
+}
+
+static void tIsr_entry_unlock(const void *p)
+{
+	k_mutex_unlock((struct k_mutex *)p);
+}
+
+static void tThread_entry_negative(void *p1, void *p2, void *p3)
+{
+	int choice = *((int *)p2);
+
+	TC_PRINT("current case is %d\n", choice);
+
+	/* Set up the fault or assert are expected before we call
+	 * the target tested funciton.
+	 */
+	switch (choice) {
+	case MUTEX_INIT_NULL:
+		set_fault_valid(true);
+		k_mutex_init(NULL);
+		break;
+	case MUTEX_LOCK_NULL:
+		set_fault_valid(true);
+		k_mutex_lock(NULL, K_NO_WAIT);
+		break;
+	case MUTEX_UNLOCK_NULL:
+		set_fault_valid(true);
+		k_mutex_unlock(NULL);
+		break;
+	case MUTEX_LOCK_IN_ISR:
+		set_assert_valid(true);
+		irq_offload(tIsr_entry_lock, (void *)p1);
+		break;
+	case MUTEX_UNLOCK_IN_ISR:
+		set_assert_valid(true);
+		irq_offload(tIsr_entry_unlock, (void *)p1);
+	default:
+		TC_PRINT("should not be here!\n");
+		break;
+	}
+
+	/* If negative comes here, it means error condition not been
+	 * detected.
+	 */
+	ztest_test_fail();
+}
+
+static int create_negative_test_thread(int choice)
+{
+	int ret;
+	uint32_t perm = K_INHERIT_PERMS;
+
+	if (_is_user_context()) {
+		perm = perm | K_USER;
+	}
+
+	case_type = choice;
+
+	k_mutex_init(&mutex);
+
+	k_tid_t tid = k_thread_create(&tdata, tstack, STACK_SIZE,
+		(k_thread_entry_t)tThread_entry_negative,
+			&mutex, (void *)&case_type, NULL,
+			K_PRIO_PREEMPT(THREAD_TEST_PRIORITY),
+			perm, K_NO_WAIT);
+
+	ret = k_thread_join(tid, K_FOREVER);
+
+	return ret;
+}
+
+/* TESTPOINT: Pass a null pointer into the API k_mutex_init */
+static void test_mutex_init_null(void)
+{
+	create_negative_test_thread(MUTEX_INIT_NULL);
+}
+/* TESTPOINT: Pass a null pointer into the API k_mutex_lock */
+static void test_mutex_lock_null(void)
+{
+	create_negative_test_thread(MUTEX_LOCK_NULL);
+}
+
+/* TESTPOINT: Pass a null pointer into the API k_mutex_unlock */
+static void test_mutex_unlock_null(void)
+{
+	create_negative_test_thread(MUTEX_UNLOCK_NULL);
+}
+
+/* TESTPOINT: Try to lock mutex in isr context */
+static void test_mutex_lock_in_isr(void)
+{
+	create_negative_test_thread(MUTEX_LOCK_IN_ISR);
+}
+
+/* TESTPOINT: Try to unlock mutex in isr context */
+static void test_mutex_unlock_in_isr(void)
+{
+	create_negative_test_thread(MUTEX_UNLOCK_IN_ISR);
+}
+
+static void test_mutex_unlock_count_unmet(void)
+{
+	struct k_mutex tmutex;
+
+	k_mutex_init(&tmutex);
+	zassert_true(k_mutex_lock(&tmutex, K_FOREVER) == 0,
+			"current thread failed to lock the mutex");
+
+	/* Verify that when current thread is equal to the lock_ower
+	 * but the lock_count is equal to zero.
+	 */
+	tmutex.lock_count = 0;
+	set_fault_valid(true);
+	k_mutex_unlock(&tmutex);
```

Comment:
My understanding is that it will causes a fatal error. My question is that how we know if a fatal error is raised?

Reply:
Actually, design of this test case is kind of a white box testing, we know there is a assertion for "count > 0" inside the k_mutex_unlock().
Strictly speaking, it should be set_assert_valid here, not set_fault_valid. It also works here because that if an assert fail happened, the normal behavior is it will go to trigger fatal error, and our fault handler still could handle it here.
And think you very much for pointing out this, we will fix it.